Add initial table with output metrics for testing

jashapiro opened the pull request:
Closes #822
Here I am adding a document with a table of metrics that will be calculated as part of testing the workflow for changes. I also included a section for formatting checks, but I have not filled that in yet. I thought we could start with just the metrics.

Most of the metrics here are derived from https://data-lab-knowledge.slab.com/posts/discussion-testing-scpca-nf-outputs-3d4pra87, with a few additions/updates. Note that some of the metrics are actually arrays or dictionaries (i.e. hvgs and  cell types). I think we do want to  include those in the output metric; they will make the file larger but still not _large_.

I wasn't sure whether we wanted to just make these part of the `_metadata.json` file or a separate `_metrics.json` file. Either way seems reasonable enough to me, with the caveat that things like the version numbers could become a bit harder to find in the metadata file if we add too many large fields to it.

Please let me know if there are any metrics that you think I missed, or if you think we should remove ones that I included.

allyhawkins commented on config/workflow-testing.md:
```diff
@@ -0,0 +1,34 @@
+# Testing scpca-nf outputs
+
+As `scpca-nf` is updated, we want to be sure that we are aware of any changes in the output from the pipeline before we deploy updated results to the portal.
+This documents outlines the tests that we will run and the key metrics that we will monitor to ensure that the pipeline is functioning as expected.
```
```suggestion
This document outlines the tests that we will run and the key metrics that we will monitor to ensure that the pipeline is functioning as expected.
```

allyhawkins commented on config/workflow-testing.md:
```diff
@@ -0,0 +1,34 @@
+This documents outlines the tests that we will run and the key metrics that we will monitor to ensure that the pipeline is functioning as expected.
+
+## Data format checks
+
```
Maybe make a note that this is coming soon or TBD?

allyhawkins commented on config/workflow-testing.md:
```diff
@@ -0,0 +1,34 @@
+## Output metrics
+
+The table below describes the key metrics that we will track for the workflow outputs, and which files are required to calculate those metrics.
+These metrics should be calculated, if required, at the end of the workflow just before the publication step, and stored in a JSON file. At the moment, I am not sure if we just add these to `_metadata.json` or create a separate file.
```
I think I favor a separate file so as not to crowd the current `_metadata.json` file? Also when we hand off the files the `_metadata.json` has information needed to deploy the samples to the Portal, but the metrics here are for internal use so I think we can keep it separate.

allyhawkins commented on config/workflow-testing.md:
```diff
@@ -0,0 +1,34 @@
+| `processed_cells`  | total number of cells in processed SCE object  | `{library_id}_metadata.json` or `{library_id}_processed.rds` |
+| `total_reads`      | total number of reads from FASTQ files | `{library_id}_metadata.json` |
+| `mapped_reads`      | total number of reads mapped by alevin-fry | `{library_id}_metadata.json` |
+| `cell_filtering_method` | method used to filter cells | `{library_id}_metadata.json` or `{library_id}_processed.rds` |
```
We should probably add in `droplet_filtering_method` and `normalization_method`.

allyhawkins commented on config/workflow-testing.md:
```diff
@@ -0,0 +1,34 @@
+| `cellassign_celltypes` | a dictionary with the number of cells in each CellAssign-based cell type  | `{library_id}_processed.rds` |
+| `unfiltered_altexp_total` | total number of reads from altExp (ADT or cellhash) | `{library_id}_unfiltered.rds` |
+| `filtered_altexp_total` | total number of reads from altExp (ADT or cellhash) | `{library_id}_filtered.rds` |
+| `processed_altexp_total` | total number of reads from altExp (ADT or cellhash) | `{library_id}_processed.rds` |
```
Do we also want to add in `adt_scpca_filter_count`, similar to the `miqc_pass_count`, but for ADT filtering?

allyhawkins reviewed (commented):
This mostly looks good, I had a few suggestions of things we should add. 
Also, I don't know that the location of this file makes the most sense? I think it probably belongs as a new header in the `internal-instructions.md` rather than in the `config` directory?

jashapiro commented:
> Also, I don't know that the location of this file makes the most sense? I think it probably belongs as a new header in the `internal-instructions.md` rather than in the `config` directory?

That's a mistake! It was meant to be top level; not sure how I ended up saving it in `config`. I didn't want it to be part of internal instructions partly because external users might want to monitor changes as well, and there will be additional files created as part of this, but I could consider making it part of that doc.

allyhawkins commented:
> > Also, I don't know that the location of this file makes the most sense? I think it probably belongs as a new header in the `internal-instructions.md` rather than in the `config` directory?
> 
> That's a mistake! It was meant to be top level; not sure how I ended up saving it in `config`. I didn't want it to be part of internal instructions partly because external users might want to monitor changes as well, and there will be additional files created as part of this, but I could consider making it part of that doc.

I think making it a separate document in the top level works for me!

jashapiro commented:
I updated the text with a bit more detail and also added a few more metrics to track. The main additions are the ones you suggestions, but also the `logcounts` matrix sum to check that for changes, and the references for SingleR and CellAssign. The thought with the last of these is that we may well want to skip checking the  assignment counts if the reference has changed, or at least to highlight that as a likely cause of any changes we might see. 

Should be ready for another look.

allyhawkins reviewed (approved):
LGTM
